### PR TITLE
Refactor CMake file and position of header files

### PR DESCRIPTION
**Pull request by therapon**
*Description of changes:*

This PR is to obtain early feedback on the approach taken to update, refactor and address #132 . 
If this is ok, then there will be another commit to address CMake packaging as well as generation of .deb and .rpm packages. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tools/events/CMakeLists.txt [hunk 1]
```diff
@@ -1,3 +1,18 @@
...
-add_library(ion_events ion_event_util.cpp ion_event_stream.cpp ion_event_equivalence.cpp)
+cmake_minimum_required(VERSION 3.5 FATAL_ERROR)
+add_library(ion_events
+        ion_event_util.cpp
+        ion_event_stream.cpp
+        ion_event_equivalence.cpp)
+
+target_include_directories(ion_events
+        PUBLIC
+            ${CMAKE_CURRENT_SOURCE_DIR}/inc
```
Comment (pbcornell) [inline tools/events/CMakeLists.txt]: should this be "include"?

File: decNumber/include/decNumber/decSingle.h [hunk 2]
```diff
@@ -38,9 +38,9 @@
   #define DECSINGLE_Ehigh (DECSINGLE_Emax + DECSINGLE_Bias - (DECSINGLE_Pmax-1))
 
   /* Required includes                                                */
-  #include "decContext.h"
-  #include "decQuad.h"
-  #include "decDouble.h"
+  #include "include/decNumber/decContext.h"
+  #include "include/decNumber/decQuad.h"
+  #include "include/decNumber/decDouble.h"
```
Comment (pbcornell) [inline decNumber/include/decNumber/decSingle.h]: change these 3 to use the `#include <decNumber/dec*.h>` form?

File: decNumber/include/decNumber/decDouble.h [hunk 3]
```diff
@@ -38,7 +38,7 @@
   #define DECDOUBLE_Ehigh (DECDOUBLE_Emax + DECDOUBLE_Bias - (DECDOUBLE_Pmax-1))
 
   /* Required includes                                                */
-  #include "decContext.h"
+  #include "include/decNumber/decContext.h"
```
Comment (pbcornell) [inline decNumber/include/decNumber/decDouble.h]: any reason not to use the following form here?

`#include <decNumber/decContext.h>`

File: decNumber/include/decNumber/decSingle.h (hunk 2 above)
Comment (therapon) [reply decNumber/include/decNumber/decSingle.h]: Yes. I thought I fixed these, apparently I missed a couple.

File: tools/events/CMakeLists.txt (hunk 1 above)
Comment (therapon) [reply tools/events/CMakeLists.txt]: This points to the subfolder for `event` which is only used internally in our code. I am not 100% sure we want this lib exposed to customers.

File: decNumber/include/decNumber/decDouble.h (hunk 3 above)
Comment (therapon) [reply decNumber/include/decNumber/decDouble.h]: ditto

**Comment by therapon** [comment]
> It's been quite a while since I've written/reviewed C, and I don't recall the expectations around include paths. That said, something seems off when I see `#include "include/..."`, which occurs a few times below.

The use of quote in includes causes the preprocessor to first search in the current folder of the file before looking in system locations (folders specified by -I). The <> use looks in system locations. 
The pattern, I believe, is to use <> for .h files that are public and "" for private to the applications code.

The `include/..` is a mistake and I will clear those out. 
> 
> The general pattern #include <_libname_/_file.h_> feels like a good approach. I assume the #include <_file.h_> form is used to include implementation logic, not APIs.
> 

Yes. The pattern is to minimize conflicts in client code by having <libname/..> prefix. 

> Regarding the move from "inc" to "include": will this break consumers?
> 

It can. It depends how IonC was consumed. The follow up PR that will address CMake packaging will make integration with IonC more modular for CMake client code (and somewhat for non-CMake client code) that will make all the build configuration (-I options) controlled by CMake.

File: tools/events/CMakeLists.txt (hunk 1 above)
Comment (tgregg) [reply tools/events/CMakeLists.txt]: Yes, let's not expose the stuff in `events` right now.

**Comment by tgregg** [comment]
The approach looks fine to me. Any idea why it's failing on Windows?

**Comment by therapon** [comment]
> The approach looks fine to me. Any idea why it's failing on Windows?

Not yet. Working on reproducing on my machine.

**Comment by tgregg** [comment]
Why do the headers to in `ionc/include/ionc` and not just `ionc/include`?

**Comment by therapon** [comment]
> Why do the headers to in `ionc/include/ionc` and not just `ionc/include`?

Mimics the Install location `/usr/local/include/ionc` and allows code within the package and code outside the package to use <ionc/include> consistently.

**Comment by therapon** [comment]
> Why do the headers to in `ionc/include/ionc` and not just `ionc/include`?

Mimics the Install location `/usr/local/include/ionc` and allows code within the package and code outside the package to use <ionc/include> consistently.

**Pull request closed without merging**